File: app/src/main/java/com/fractalwrench/crazycats/resource/JsonResourceReader.java
package com.fractalwrench.crazycats.resource;

import com.squareup.moshi.Moshi;

import java.io.IOException;

/**
 * Reads a resource file and serialises it from a JSON String.
 * @param <T> the type of class to serialise into
 */
public class JsonResourceReader<T> {

    private final StringResourceFileReader resourceFileReader = new StringResourceFileReader();
    private final Moshi moshi;
    private final Class<T> clz;

    public JsonResourceReader(Moshi moshi, Class<T> clz) {
        this.moshi = moshi;
        this.clz = clz;
    }

    public T readResourceAsJson(String resName) throws IOException {
        String json = resourceFileReader.readResourceAsString(resName);
        return moshi.adapter(clz).fromJson(json);
    }

}

File: app/src/main/java/com/fractalwrench/crazycats/resource/JsonResourceReader.kt
package com.fractalwrench.crazycats.resource

import com.squareup.moshi.Moshi

import java.io.IOException

/**
 * Reads a resource file and serialises it from a JSON String.
 * @param <T> the type of class to serialise into
</T> */
class JsonResourceReader<T>(private val moshi: Moshi, private val clz: Class<T>) {

    private val resourceFileReader = StringResourceFileReader()

    @Throws(IOException::class)
    fun readResourceAsJson(resName: String): T {
        val json = resourceFileReader.readResourceAsString(resName)
        return moshi.adapter(clz).fromJson(json)
    }

}
